platform/mellanox/docker-syncd-mlnx/ecmp_calc_sdk.py

```diff
@@ -0,0 +1,466 @@
+    
+    Args:
+        handle (sx_api_handle_t): SDK handle
+        vport (sx_port_id_t): SDK vport id
+
+    Returns:
+        sx_port_log_id_t : SDK logical index 
+    """      
+    try:
+        vlan_id_p = new_sx_vlan_id_t_p()
+        logical_port_p = new_sx_port_log_id_t_p()
+        
+        rc = sx_api_port_vport_base_get(handle, vport, vlan_id_p, logical_port_p)
+        sx_check_rc(rc)
+        
+        logical_port = sx_port_log_id_t_p_value(logical_port_p)
+        
+        return logical_port  
+      
+    finally:
+        delete_sx_vlan_id_t_p(vlan_id_p)
+        delete_sx_port_log_id_t_p(logical_port_p)   
+        
+def sx_router_neigh_get_mac(handle, rif, addr, type_ipv6=False):
+    """ Get neighbour mac address
+    
+    Args:
+        handle (sx_api_handle_t): SDK handle
+        rif (sx_port_id_t): SDK vport id
+        addr (sx_ip_addr_t): Neighbour IP address
+        type_ipv6 (bool): True for IPv6
+
+    Returns:
+        str : Neighbour mac address
+    """     
+    try:
+        neigh_entry_cnt_p = new_uint32_t_p()
+        neigh_entry_list_p = new_sx_neigh_get_entry_t_arr(1)
+
+        filter_p = new_sx_neigh_filter_t_p()
+        neigh_filter = sx_neigh_filter_t()
+        neigh_filter.filter_by_rif = SX_KEY_FILTER_FIELD_NOT_VALID
+        neigh_filter.rif = 0
+        sx_neigh_filter_t_p_assign(filter_p, neigh_filter)
+        
+        rc = sx_api_router_neigh_get(handle, SX_ACCESS_CMD_GET, rif, addr, filter_p, neigh_entry_list_p, neigh_entry_cnt_p)
+        sx_check_rc(rc)
+
+        neighbor_entry = sx_neigh_get_entry_t_arr_getitem(neigh_entry_list_p, 0)
+        
+        return neighbor_entry.neigh_data.mac_addr.to_str()
+    
+    finally:
+        delete_sx_neigh_filter_t_p(filter_p)
+        delete_uint32_t_p(neigh_entry_cnt_p)
+        delete_sx_neigh_get_entry_t_arr(neigh_entry_list_p)    
+        
+        
+def sx_fdb_uc_mac_addr_get(handle, vlan_id, mac_addr):
+    """ Get UC mac entry from FDB
+    
+    Args:
+        handle (sx_api_handle_t): SDK handle
+        vlan_id (sx_vlan_id_t): VLAN id
+        mac_addr (str): mac address
+
+    Returns:
+        sx_fdb_uc_mac_addr_params_t : FDB mac entry
+    """       
+    try:
+        key = sx_fdb_uc_mac_addr_params_t()
+        key.fid_vid = vlan_id
+        key.mac_addr = ether_addr(mac_addr)
+        key.entry_type = SX_FDB_UC_AGEABLE
+        key.action = SX_FDB_ACTION_FORWARD
+        key_p = copy_sx_fdb_uc_mac_addr_params_t_p(key)
+        
+        key_filter = sx_fdb_uc_key_filter_t()
+        key_filter.filter_by_fid =  SX_FDB_KEY_FILTER_FIELD_VALID
+        key_filter.filter_by_mac_addr = SX_FDB_KEY_FILTER_FIELD_VALID
+        key_filter.filter_by_log_port = SX_FDB_KEY_FILTER_FIELD_NOT_VALID
+        key_filter.fid = vlan_id
+        key_filter.mac_addr = ether_addr(mac_addr)
+        key_filter_p = copy_sx_fdb_uc_key_filter_t_p(key_filter)
+    
+        data_cnt_p = copy_uint32_t_p(SX_FDB_MAX_GET_ENTRIES)
```

Comment:
@liorghub are we sure that this call does allocate memory and not simply makes a pointer?

Reply:
Yes, it is allocating memory. It is implemented in sonic-sairedis/pyext/py2/pysairedis_wrap.cpp.
Function _wrap_copy_uint32_t_p is calling SWIG_NewPointerObj which does the allocation.